src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -282,21 +282,26 @@ abstract class Erasure extends InfoTransform
             }
           def classSig = {
             val preRebound = pre.baseType(sym.owner) // #2585
-            dotCleanup(
-              (
-                if (needsJavaSig(preRebound, Nil)) {
-                  val s = jsig(preRebound, existentiallyBound)
-                  if (s.charAt(0) == 'L') s.substring(0, s.length - 1) + "." + sym.javaSimpleName
-                  else fullNameInSig(sym)
-                }
-                else fullNameInSig(sym)
-              ) + (
-                if (args.isEmpty) "" else
-                "<"+(args map argSig).mkString+">"
-              ) + (
-                ";"
-              )
-            )
+            val sigCls = {
+              if (needsJavaSig(preRebound, Nil)) {
+                val s = jsig(preRebound, existentiallyBound)
+                if (s.charAt(0) == 'L') {
+                  val withoutSemi = s.substring(0, s.length - 1)
+                  // If the prefix is a module, drop the '$'. Classes (or modules) nested in modules
+                  // are separated by a single '$' in the filename: `object o { object i }` is o$i$.
+                  val withoutOwningModuleDollar =
+                    if (preRebound.typeSymbol.isModuleClass) withoutSemi.stripSuffix(nme.MODULE_SUFFIX_STRING)
```

Comment:
In your version (https://github.com/scala/scala/pull/5690/commits/21b4378a04d404a2e2c7547b28ce2ee3d7f268fe#diff-34082ec02c067748f0d47c0bf037a717R292) there's an additional check `&& sym.isModuleOrModuleClass`, but that's not correct: we also need to do it for classes, see the last line in the test of this commit.